src/compiler/ir/README.md

```diff
@@ -0,0 +1,522 @@
+This requirement applies regardless of the kind of value held by the variable:
+even an integer value must be the subject of a destructive use, albeit this
+will lower to a no-op.
+
+A destructive use is not considered a regular use. Therefore, a variable may
+(and must) be destroyed after a consuming use. This implies that `%x = ...; %y
+= call f(%x)` on its own is not a valid IR. It should instead have the form `%x
+= ...; %y = call f(%x); end-scope(%x)`. If `%x` holds an owned value, the
+consuming use in the call to `f` will have moved the value. The `end-scope`
+statement is therefore a no-op.
+
+## Forwarding uses
+
+When a variable is passed as an argument to another basic block, we should
+treat this use differently based on how the target basic block uses its
+parameter. For this reason, we consider arguments passed to basic blocks
+**forwarding uses**.
+
+Consider the following IR:
+```
+BB0:
+  %x0 = ...
+  %y0 = ...
+  goto BB1(%x0, %y0)
+
+BB1(%x1, %y1):
+  use(%1)  // Regular use of %1
+  end-scope(%x1, %y1)
+```
+
+In this example, `%x1` is the subject of a regular use, whereas `%y1` is not.
+Therefore, due to the forwarding nature of the `goto` statement, `%x0` is the
+subject of a regular use but `%y0` is not.
+
+Because all variables must eventually be the subject of a destructive use, here
+the `end-scope(%x1, %y1)` statement, forwarding uses are themselves always
+considered destructive uses. This has an important consequence: if a variable
+is passed as basic block argument, it may not be used again. In the example
+above, it would be illegal to use `%x0` or `%y0` in `BB1`. Additionally, such a
+variable does not need any additional destructive use: note the absence of a
+`end-scope(%x0, %y0)`.
+
+TODO: we actually use the word "destructive use" for two things, `(overwrite +
+end-scope)` and `(overwrite + end-scope + block arguments)`. The latter
+includes forwarding uses, since these "act like" a destructive use in many
+aspects.
+
+TODO: Should variable uses be evaluated as a least or greatest fixed point?. In
+other words, in `BB0(%x): goto BB0(%x)`, is `%x` ever the subject of a regular,
+consuming or destructive use?  While an interesting theoretical question, I
+don't think we need to worry too much about this in practice.
+
+# Conditionally moved values
+
+As mentioned earlier, if a value is moved, its destructive use has no effect.
+Determining at the destructive use site whether or not to release a value may
+require tracking additional information at runtime, in at least two ways.
+
+Firstly, a consuming use may have occurred inside a branch. The value would
+have been moved only if the branch is taken. In this case, the final
+`end-scope` effect should only release the value was not taken.
+
+```
+BB0:
+  %x = ...
+  %y = ... // %y holds an owned value
+  if %x then goto BB1 else goto BB2
+
+BB1:
+  call foo(%y) // Consuming use of %y, its value is moved
```

Comment:
I would have thought the `end-scope(%y)` would be here, right after use, not after the PHI node in the tail block. This would get rid of the conundrum, no?

Reply:
`end-scope` represents the point in the surface language where a variable goes out of scope, ie. the block in which that variable was declared in closed.

In this particular example, nothing happens between the call and the end-scope. However a more complex example:

```pony
{
  var y = ...;
  if * {
    foo(y); // y is conditionally moved
  }
  bar();
  // y goes out of scope
}
```

the `end-scope` cannot be simply moved into the conditional block.